mapping/field_dv.go

```diff
@@ -0,0 +1,203 @@
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"github.com/blevesearch/bleve/v2/geo"
+	"github.com/blevesearch/bleve/v2/document"
+)
+
+// A FieldMapping describes how a specific item
+// should be put into the index.
+type FieldMapping struct {
+	Name string `json:"name,omitempty"`
+	Type string `json:"type,omitempty"`
+
+	// Analyzer specifies the name of the analyzer to use for this field.  If
+	// Analyzer is empty, traverse the DocumentMapping tree toward the root and
+	// pick the first non-empty DefaultAnalyzer found. If there is none, use
+	// the IndexMapping.DefaultAnalyzer.
+	Analyzer string `json:"analyzer,omitempty"`
+
+	// Store indicates whether to store field values in the index. Stored
+	// values can be retrieved from search results using SearchRequest.Fields.
+	Store bool `json:"store,omitempty"`
+	Index bool `json:"index,omitempty"`
+
+	// IncludeTermVectors, if true, makes terms occurrences to be recorded for
+	// this field. It includes the term position within the terms sequence and
+	// the term offsets in the source document field. Term vectors are required
+	// to perform phrase queries or terms highlighting in source documents.
+	IncludeTermVectors bool   `json:"include_term_vectors,omitempty"`
+	IncludeInAll       bool   `json:"include_in_all,omitempty"`
+	DateFormat         string `json:"date_format,omitempty"`
+
+	// DocValues, if true makes the index uninverting possible for this field
+	// It is useful for faceting and sorting queries.
+	DocValues bool `json:"docvalues,omitempty"`
+
+	// SkipFreqNorm, if true, avoids the indexing of frequency and norm values
+	// of the tokens for this field. This option would be useful for saving
+	// the processing of freq/norm details when the default score based relevancy
+	// isn't needed.
+	SkipFreqNorm bool `json:"skip_freq_norm,omitempty"`
+
+	// Dimension of the dense vector
+	Dims int `json:"dims,omitempty"`
+
+	// Similarity is the similarity algorithm used for scoring
+	// dense_vector fields. See also SimilarityDefault and SimilarityValid
+	Similarity string `json:"similarity,omitempty"`
+}
+
+// Default Similarity value for dense_vector fields
+const SimilarityDefault = "l2_norm"
+
+// Set of Valid Similarity values for dense_vector fields
+var SimilarityValid = map[string]bool{
+	"l2_norm":     true, // default
+	"cosine":      false,
+	"dot_product": false,
+}
+
+func NewDenseVectorFieldMapping() *FieldMapping {
+	return &FieldMapping{
+		Type:         "densevector",
+		Store:        false,
+		Index:        true,
+		IncludeInAll: false,
+		DocValues:    false,
+	}
+}
+
+func (fm *FieldMapping) processDenseVector(propertyMightBeDenseVector interface{},
+	pathString string, path []string, indexes []uint64, context *walkContext) {
+	propertyVal := reflect.ValueOf(propertyMightBeDenseVector)
+	if !propertyVal.IsValid() {
+		return
+	}
+
+	if propertyVal.Kind() == reflect.Slice {
+		dv := make([]float32, propertyVal.Len())
+		for i := 0; i < propertyVal.Len(); i++ {
```

Comment:
can we avoid this loop and directly convert Slice to []float32

Reply:
We can't avoid the loop here, because,

Essentially we have a []interface{}  and we want to check if it's a []float32.

An []interface{} can take values like { 1.1, 2, "shaad"}       ( values of heterogenous types )
Thus, without going through each element of the slice we can't say that it is a slice of float32s.



#### One may tempt to do a type assertion, like 
```
x : []interface{}{1, 2, "shaad"}
if y, ok := x.([]float32); ok { 
 ...
}
```
But it won't work as x is of type []interface{} and not of type []float32

#### There is poorly performant way of avoiding for loop ( but it kills the purpose ) 

```
 x := []interface{}{1, 2, 3.2}
 y := []float64{}
 b, err := json.Marshal(x)
 if err == nil {
   err = json.Unmarshal(b, &y)
   if err == nil {
     fmt.Println("y:", y)
   }
 }
```
For a vector of size 1000, this approach is 12 times worse than loop approach :)